File: python/cmy.py
from concurrent.futures import ProcessPoolExecutor
import time
from PIL import Image
import numpy as np
import cupy as cp

THREADS = 16

WRITE_FILE = False
FILE = 'big.jpg'
OUT_FILE = 'big_out.jpg'

def rgb_to_cmy(rgb:tuple[int,int,int]) -> tuple[int,int,int]:
    out = np.array([
        255 - rgb[0],
        255 - rgb[1],
        255 - rgb[2]
    ], dtype=np.uint8)
    return out


def cpu_serial():
    im: Image = Image.open(FILE)
    im_np = np.array(im)

    height, width , _ = im_np.shape
    
    im_np_out = np.zeros((height, width, 3), dtype=np.uint8)

    for i in range(height):
        for j in range(width):
            im_np_out[i][j] = rgb_to_cmy(im_np[i][j])

    im_out = Image.fromarray(im_np_out)
    if (WRITE_FILE):
        im_out.save(OUT_FILE)
    
    

def _process_rows(rows: np.ndarray) -> np.ndarray:
    out = np.zeros(rows.shape[:2] + (3,), dtype=np.uint8)
    for i in range(rows.shape[0]):
        for j in range(rows.shape[1]):
            out[i][j] = rgb_to_cmy(rows[i][j])
    return out


def cpu_parallel():
    im: Image = Image.open(FILE)
    im_np = np.array(im)

    num_workers = THREADS
    chunks = np.array_split(im_np, num_workers, axis=0)

    with ProcessPoolExecutor(max_workers=num_workers) as executor:
        results = list(executor.map(_process_rows, chunks))

    im_np_out = np.concatenate(results, axis=0)

    im_out = Image.fromarray(im_np_out)
    if (WRITE_FILE):
        im_out.save(OUT_FILE)

def _process_rows_gpu_wrong(rows: np.ndarray) -> np.ndarray:
    rows_gpu = cp.asarray(rows)
    rows_cpu = cp.asnumpy(rows_gpu)

    out = np.zeros(rows_cpu.shape[:2] + (3,), dtype=np.uint8)
    for i in range(rows_cpu.shape[0]):
        for j in range(rows_cpu.shape[1]):
            out[i][j] = rgb_to_cmy(rows_cpu[i][j])
    return out


def gpu_parallel_wrong():
    im: Image = Image.open(FILE)
    im_np = np.array(im)

    num_workers = THREADS
    chunks = np.array_split(im_np, num_workers, axis=0)

    with ProcessPoolExecutor(max_workers=num_workers) as executor:
        results = list(executor.map(_process_rows_gpu_wrong, chunks))

    im_np_out = np.concatenate(results, axis=0)

    im_out = Image.fromarray(im_np_out)
    if (WRITE_FILE):
        im_out.save(OUT_FILE)

def gpu_parallel_right():
    im: Image = Image.open(FILE)
    im_np = np.array(im)

    im_gpu = cp.asarray(im_np)
    im_gpu_out = 255 - im_gpu
    im_np_out = cp.asnumpy(im_gpu_out)

    im_out = Image.fromarray(im_np_out)
    if (WRITE_FILE):
        im_out.save(OUT_FILE)  


def main():

    start = time.perf_counter()
    cpu_serial()
    print(f"CPU serial: {time.perf_counter() - start:.4f}s")

    start = time.perf_counter()
    cpu_parallel()
    print(f"CPU parallel: {time.perf_counter() - start:.4f}s")

    start = time.perf_counter()
    gpu_parallel_wrong()
    print(f"GPU parallel (wrong): {time.perf_counter() - start:.4f}s")

    start = time.perf_counter()
    gpu_parallel_right()
    print(f"GPU parallel (right): {time.perf_counter() - start:.4f}s")


if __name__ == "__main__":
    main()

File: julia/cmy.jl
using Images
using FileIO
using CUDA

const THREADS = 16

const WRITE_FILE = true
const FILE = joinpath(@__DIR__, "..", "big.jpg")
const OUT_FILE = joinpath(@__DIR__, "..", "big_out.jpg")

rgb_to_cmy(r::UInt8, g::UInt8, b::UInt8) = (0xff - r, 0xff - g, 0xff - b)

function load_raw(path)
    img = load(path)                          # H x W Array{RGB{N0f8}}
    return collect(rawview(channelview(img)))  # materialized 3 x H x W UInt8 array
end

function save_raw(raw::AbstractArray{UInt8,3}, path)
    img = colorview(RGB, normedview(raw))
    save(path, img)
end

function cpu_serial()
    raw = load_raw(FILE)
    _, height, width = size(raw)
    out = Array{UInt8}(undef, size(raw))

    for i in 1:height
        for j in 1:width
            out[1, i, j], out[2, i, j], out[3, i, j] =
                rgb_to_cmy(raw[1, i, j], raw[2, i, j], raw[3, i, j])
        end
    end

    WRITE_FILE && save_raw(out, OUT_FILE)
end

function _process_rows!(out, raw, rows)
    for i in rows
        for j in 1:size(raw, 3)
            out[1, i, j], out[2, i, j], out[3, i, j] =
                rgb_to_cmy(raw[1, i, j], raw[2, i, j], raw[3, i, j])
        end
    end
end

function cpu_parallel()
    raw = load_raw(FILE)
    _, height, width = size(raw)
    out = Array{UInt8}(undef, size(raw))

    num_workers = min(THREADS, Threads.nthreads())
    chunks = Iterators.partition(1:height, cld(height, num_workers))

    Threads.@threads for rows in collect(chunks)
        _process_rows!(out, raw, rows)
    end

    WRITE_FILE && save_raw(out, OUT_FILE)
end

# GPU "wrong": ships each row chunk to the GPU and immediately back, then
# still does the actual conversion element-by-element on the CPU. The
# transfer overhead dominates and the GPU does no real work.
function _process_rows_gpu_wrong!(out, raw, rows)
    chunk_cpu = Array(CuArray(raw[:, rows, :]))
    for (k, i) in enumerate(rows)
        for j in 1:size(raw, 3)
            out[1, i, j], out[2, i, j], out[3, i, j] =
                rgb_to_cmy(chunk_cpu[1, k, j], chunk_cpu[2, k, j], chunk_cpu[3, k, j])
        end
    end
end

function gpu_parallel_wrong()
    raw = load_raw(FILE)
    _, height, width = size(raw)
    out = Array{UInt8}(undef, size(raw))

    num_workers = min(THREADS, Threads.nthreads())
    chunks = Iterators.partition(1:height, cld(height, num_workers))

    Threads.@threads for rows in collect(chunks)
        _process_rows_gpu_wrong!(out, raw, rows)
    end

    WRITE_FILE && save_raw(out, OUT_FILE)
end

function gpu_parallel_right()
    raw = load_raw(FILE)

    raw_gpu = CuArray(raw)
    out_gpu = 0xff .- raw_gpu
    out = Array(out_gpu)

    WRITE_FILE && save_raw(out, OUT_FILE)
end

function warmup_gpu()
    dummy = CuArray(UInt8[0, 0])
    Array(0xff .- dummy)
    CUDA.synchronize()
end

function main()
    warmup_gpu() # forces one-time kernel JIT compilation before timing starts

    start = time()
    cpu_serial()
    println("CPU serial: $(round(time() - start, digits=4))s")

    start = time()
    cpu_parallel()
    println("CPU parallel ($(min(THREADS, Threads.nthreads())) threads): $(round(time() - start, digits=4))s")

    start = time()
    gpu_parallel_wrong()
    println("GPU parallel (wrong): $(round(time() - start, digits=4))s")

    start = time()
    gpu_parallel_right()
    println("GPU parallel (right): $(round(time() - start, digits=4))s")
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
